File: src/main/java/net/ccbluex/liquidbounce/features/module/modules/player/NoFall.java
/*
 * FDPClient Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge by LiquidBounce.
 * https://github.com/UnlegitMC/FDPClient/
 */
package net.ccbluex.liquidbounce.features.module.modules.player;

import net.ccbluex.liquidbounce.LiquidBounce;
import net.ccbluex.liquidbounce.event.*;
import net.ccbluex.liquidbounce.features.module.Module;
import net.ccbluex.liquidbounce.features.module.ModuleCategory;
import net.ccbluex.liquidbounce.features.module.ModuleInfo;
import net.ccbluex.liquidbounce.features.module.modules.movement.Fly;
import net.ccbluex.liquidbounce.features.module.modules.render.FreeCam;
import net.ccbluex.liquidbounce.utils.PacketUtils;
import net.ccbluex.liquidbounce.utils.RotationUtils;
import net.ccbluex.liquidbounce.utils.VecRotation;
import net.ccbluex.liquidbounce.utils.MovementUtils;
import net.ccbluex.liquidbounce.utils.block.BlockUtils;
import net.ccbluex.liquidbounce.utils.misc.FallingPlayer;
import net.ccbluex.liquidbounce.utils.timer.TickTimer;
import net.ccbluex.liquidbounce.value.IntegerValue;
import net.ccbluex.liquidbounce.value.FloatValue;
import net.ccbluex.liquidbounce.value.BoolValue;
import net.ccbluex.liquidbounce.value.ListValue;
import net.minecraft.block.Block;
import net.minecraft.block.BlockLiquid;
import net.minecraft.network.play.client.C03PacketPlayer;
import net.minecraft.network.play.client.C08PacketPlayerBlockPlacement;
import net.minecraft.network.play.client.C09PacketHeldItemChange;
import net.minecraft.item.*;
import net.minecraft.util.*;
import net.minecraft.init.Blocks;
import net.minecraft.init.Items;

import java.util.ArrayList;
import java.util.Timer;
import java.util.TimerTask;

@ModuleInfo(name = "NoFall", category = ModuleCategory.PLAYER)
public class NoFall extends Module {
    public final ListValue modeValue = new ListValue("Mode", new String[]{"SpoofGround", "AlwaysSpoofGround", "NoGround", "Packet", "Packet1", "MLG", "OldAAC", "LAAC", "AAC3.3.11", "AAC3.3.15", "AACv4", "AAC5.0.14", "Spartan", "CubeCraft", "Hypixel","HypSpoof","Phase", "Verus", "Damage"}, "SpoofGround");
    private final FloatValue minFallDistance = new FloatValue("MinMLGHeight", 5F, 2F, 50F);
    private final BoolValue voidCheck = new BoolValue("Void-Check", true);
    private final IntegerValue phaseOffsetValue = (IntegerValue) new IntegerValue("PhaseOffset",1,0,5).displayable(() -> modeValue.equals("Phase"));

    private int state;
    private boolean jumped;
    private final TickTimer spartanTimer = new TickTimer();
    private final TickTimer mlgTimer = new TickTimer();
    private VecRotation currentMlgRotation;
    private int currentMlgItemIndex;
    private BlockPos currentMlgBlock;

    private boolean aac4Fakelag=false;
    private boolean packetModify =false;
    private boolean aac5doFlag=false;
    private boolean aac5Check=false;
    private int aac5Timer=0;
    private final ArrayList<C03PacketPlayer> aac4Packets=new ArrayList<>();
    private boolean needSpoof=false;
    private int packet1Count=0;

    @Override
    public void onEnable(){
        aac4Fakelag=false;
        aac5Check=false;
        packetModify =false;
        aac4Packets.clear();
        needSpoof =false;
        aac5doFlag=false;
        aac5Timer=0;
        packet1Count=0;
    }

    @EventTarget(ignoreCondition = true)
    public void onUpdate(UpdateEvent event) {
        if(mc.thePlayer.onGround)
            jumped = false;

        if(mc.thePlayer.motionY > 0)
            jumped = true;

        if (!getState() || LiquidBounce.moduleManager.getModule(FreeCam.class).getState())
            return;

        if (mc.thePlayer.isSpectator() || mc.thePlayer.capabilities.allowFlying || mc.thePlayer.capabilities.disableDamage)
            return;

        if(BlockUtils.collideBlock(mc.thePlayer.getEntityBoundingBox(), block -> block instanceof BlockLiquid) ||
                BlockUtils.collideBlock(new AxisAlignedBB(mc.thePlayer.getEntityBoundingBox().maxX, mc.thePlayer.getEntityBoundingBox().maxY, mc.thePlayer.getEntityBoundingBox().maxZ, mc.thePlayer.getEntityBoundingBox().minX, mc.thePlayer.getEntityBoundingBox().minY - 0.01D, mc.thePlayer.getEntityBoundingBox().minZ), block -> block instanceof BlockLiquid))
            return;
        
        if (!LiquidBounce.moduleManager.getModule(Fly.class).getState() && voidCheck.get() && inVoid()) return;

        switch(modeValue.get().toLowerCase()) {
            case "packet": {
                if (mc.thePlayer.fallDistance > 2F)
                    mc.getNetHandler().addToSendQueue(new C03PacketPlayer(true));
                break;
            }
            case "cubecraft": {
                if (mc.thePlayer.fallDistance > 2F) {
                    mc.thePlayer.onGround = false;
                    mc.getNetHandler().addToSendQueue(new C03PacketPlayer(true));
                }
                break;
            }
            case "oldaac": {
                if (mc.thePlayer.fallDistance > 2F) {
                    mc.getNetHandler().addToSendQueue(new C03PacketPlayer(true));
                    state = 2;
                } else if (state == 2 && mc.thePlayer.fallDistance < 2) {
                    mc.thePlayer.motionY = 0.1D;
                    state = 3;
                    return;
                }

                switch (state) {
                    case 3:
                        mc.thePlayer.motionY = 0.1D;
                        state = 4;
                        break;
                    case 4:
                        mc.thePlayer.motionY = 0.1D;
                        state = 5;
                        break;
                    case 5:
                        mc.thePlayer.motionY = 0.1D;
                        state = 1;
                        break;
                }
                break;
            }
            case "laac": {
                if (!jumped && mc.thePlayer.onGround && !mc.thePlayer.isOnLadder() && !mc.thePlayer.isInWater()
                        && !mc.thePlayer.isInWeb)
                    mc.thePlayer.motionY = -6;
                break;
            }
            case "aac3.3.11": {
                if (mc.thePlayer.fallDistance > 2) {
                    mc.thePlayer.motionX = mc.thePlayer.motionZ = 0;
                    mc.getNetHandler().addToSendQueue(new C03PacketPlayer.C04PacketPlayerPosition(mc.thePlayer.posX,
                            mc.thePlayer.posY - 10E-4D, mc.thePlayer.posZ, mc.thePlayer.onGround));
                    mc.getNetHandler().addToSendQueue(new C03PacketPlayer(true));
                }
                break;
            }
            case "aac3.3.15": {
                if (mc.thePlayer.fallDistance > 2) {
                    if (!mc.isIntegratedServerRunning())
                        mc.getNetHandler().addToSendQueue(new C03PacketPlayer.C04PacketPlayerPosition(mc.thePlayer.posX,
                                Double.NaN, mc.thePlayer.posZ, false));
                    mc.thePlayer.fallDistance = -9999;
                }
                break;
            }
            case "spartan": {
                spartanTimer.update();

                if (mc.thePlayer.fallDistance > 1.5 && spartanTimer.hasTimePassed(10)) {
                    mc.getNetHandler().addToSendQueue(new C03PacketPlayer.C04PacketPlayerPosition(mc.thePlayer.posX,
                            mc.thePlayer.posY + 10, mc.thePlayer.posZ, true));
                    mc.getNetHandler().addToSendQueue(new C03PacketPlayer.C04PacketPlayerPosition(mc.thePlayer.posX,
                            mc.thePlayer.posY - 10, mc.thePlayer.posZ, true));
                    spartanTimer.reset();
                }
                break;
            }
            case "aac5.0.14": {
                double offsetYs = 0.0;
                aac5Check=false;
                while (mc.thePlayer.motionY-1.5 < offsetYs) {
                    final BlockPos blockPos = new BlockPos(mc.thePlayer.posX, mc.thePlayer.posY + offsetYs, mc.thePlayer.posZ);
                    final Block block = BlockUtils.getBlock(blockPos);
                    final AxisAlignedBB axisAlignedBB = block.getCollisionBoundingBox(mc.theWorld, blockPos, BlockUtils.getState(blockPos));
                    if(axisAlignedBB != null) {
                        offsetYs = -999.9;
                        aac5Check=true;
                    }
                    offsetYs -= 0.5;
                }
                if(mc.thePlayer.onGround) {
                    mc.thePlayer.fallDistance = -2;
                    aac5Check=false;
                }
                if(aac5Timer>0) {
                    aac5Timer -= 1;
                }
                if(aac5Check && mc.thePlayer.fallDistance>2.5 && !mc.thePlayer.onGround) {
                    aac5doFlag = true;
                    aac5Timer = 19;
                }else {
                    if(aac5Timer<2) aac5doFlag=false;
                }
                if(aac5doFlag) {
                    if(mc.thePlayer.onGround) {
                        mc.getNetHandler().addToSendQueue(new C03PacketPlayer.C04PacketPlayerPosition(mc.thePlayer.posX,
                                mc.thePlayer.posY + 0.5, mc.thePlayer.posZ, true));
                    }else{
                        mc.getNetHandler().addToSendQueue(new C03PacketPlayer.C04PacketPlayerPosition(mc.thePlayer.posX,
                                mc.thePlayer.posY + 0.42, mc.thePlayer.posZ, true));
                    }
                }
                break;
            }
            case "phase":{
                if(mc.thePlayer.fallDistance > (3+phaseOffsetValue.get())){
                    BlockPos fallPos=new FallingPlayer(mc.thePlayer.posX,mc.thePlayer.posY,mc.thePlayer.posZ,mc.thePlayer.motionX,mc.thePlayer.motionY,mc.thePlayer.motionZ,mc.thePlayer.rotationYaw,mc.thePlayer.moveStrafing,mc.thePlayer.moveForward)
                            .findCollision(5);
                    if(fallPos==null)
                        return;

                    if((fallPos.getY()-(mc.thePlayer.motionY/20.0))<mc.thePlayer.posY){
                        mc.timer.timerSpeed=0.05F;
                        new Timer().schedule(new TimerTask() {
                            @Override
                            public void run() {
                                mc.getNetHandler().addToSendQueue(new C03PacketPlayer.C04PacketPlayerPosition(fallPos.getX(),fallPos.getY(),fallPos.getZ(),true));
                                mc.timer.timerSpeed=1F;
                            }
                        },100);
                    }
                }
            }
            case "verus":{
                if(mc.thePlayer.fallDistance - mc.thePlayer.motionY > 3) {
                    mc.thePlayer.motionY = 0.0;
                    mc.thePlayer.fallDistance = 0.0f;
                    mc.thePlayer.motionX *= 0.6;
                    mc.thePlayer.motionZ *= 0.6;
                    needSpoof = true;
                }
            }
            case "packet1":{
                if((((int)mc.thePlayer.fallDistance)/3)>packet1Count){
                    packet1Count=((int)mc.thePlayer.fallDistance)/3;
                    packetModify=true;
                }
                if(mc.thePlayer.onGround){
                    packet1Count=0;
                }
            }
        }
    }

    @EventTarget
    public void onMotion(final MotionEvent event){
        if(modeValue.equals("AACv4")&&event.getEventState() == EventState.PRE){
            if (!inVoid()) {
                if (aac4Fakelag) {
                    aac4Fakelag = false;
                    if (aac4Packets.size() > 0) {
                        for (C03PacketPlayer packet : aac4Packets) {
                            mc.thePlayer.sendQueue.addToSendQueue(packet);
                        }
                        aac4Packets.clear();
                    }
                }
                return;
            }
            if (mc.thePlayer.onGround && aac4Fakelag) {
                aac4Fakelag = false;
                if (aac4Packets.size() > 0) {
                    for (C03PacketPlayer packet : aac4Packets) {
                        mc.thePlayer.sendQueue.addToSendQueue(packet);
                    }
                    aac4Packets.clear();
                }
                return;
            }
            if (mc.thePlayer.fallDistance > 2.5 && aac4Fakelag) {
                packetModify = true;
                mc.thePlayer.fallDistance = 0;
            }
            if (inAir(4.0, 1.0)) {
                return;
            }
            if (!aac4Fakelag) {
                aac4Fakelag = true;
            }
        }
    }

    @EventTarget
    public void onPacket(final PacketEvent event) {
        final String mode = modeValue.get();

        if (mc.thePlayer == null) return;
        if (!LiquidBounce.moduleManager.getModule(Fly.class).getState() && voidCheck.get() && inVoid()) return;

        if (event.getPacket() instanceof C03PacketPlayer) {
            final C03PacketPlayer packet = (C03PacketPlayer) event.getPacket();

            if (mode.equalsIgnoreCase("SpoofGround") && mc.thePlayer.fallDistance>2.5)
                packet.onGround = true;

            if (mode.equalsIgnoreCase("AlwaysSpoofGround"))
                packet.onGround = true;

            if (mode.equalsIgnoreCase("NoGround"))
                packet.onGround = false;

            if (mode.equalsIgnoreCase("Hypixel")
                    && mc.thePlayer != null && mc.thePlayer.fallDistance > 1.5)
                packet.onGround = mc.thePlayer.ticksExisted % 2 == 0;

            if (mode.equalsIgnoreCase("HypSpoof"))
                PacketUtils.sendPacketNoEvent(new C03PacketPlayer.C04PacketPlayerPosition(packet.x,packet.y,packet.z,true));

            if (mode.equalsIgnoreCase("AACv4")&&aac4Fakelag){
                event.cancelEvent();
                if (packetModify) {
                    packet.onGround = true;
                    packetModify = false;
                }
                aac4Packets.add(packet);
            }

            if (mode.equalsIgnoreCase("Verus") && needSpoof) {
                packet.onGround = true;
                needSpoof = false;
            }
            if (mode.equalsIgnoreCase("Damage") && mc.thePlayer != null && mc.thePlayer.fallDistance > 3.5) 
                packet.onGround = true;
            
            if (mode.equalsIgnoreCase("Packet1") && packetModify) {
                packet.onGround = true;
                packetModify = false;
            }
        }
    }

    @EventTarget
    public void onMove(MoveEvent event) {
        if(BlockUtils.collideBlock(mc.thePlayer.getEntityBoundingBox(), block -> block instanceof BlockLiquid) || BlockUtils.collideBlock(new AxisAlignedBB(mc.thePlayer.getEntityBoundingBox().maxX, mc.thePlayer.getEntityBoundingBox().maxY, mc.thePlayer.getEntityBoundingBox().maxZ, mc.thePlayer.getEntityBoundingBox().minX, mc.thePlayer.getEntityBoundingBox().minY - 0.01D, mc.thePlayer.getEntityBoundingBox().minZ), block -> block instanceof BlockLiquid))
            return;
        
        if (!LiquidBounce.moduleManager.getModule(Fly.class).getState() && voidCheck.get() && inVoid()) return;

        if (modeValue.equals("laac")) {
            if (!jumped && !mc.thePlayer.onGround && !mc.thePlayer.isOnLadder() && !mc.thePlayer.isInWater() && !mc.thePlayer.isInWeb && mc.thePlayer.motionY < 0D) {
                event.setX(0);
                event.setZ(0);
            }
        }
    }

    @EventTarget(ignoreCondition = true)
    public void onJump(final JumpEvent event) {
        jumped = true;
    }

    private boolean inVoid(){
        if (mc.thePlayer.posY < 0) {
            return false;
        }
        for (int off = 0; off < mc.thePlayer.posY + 2; off += 2) {
            AxisAlignedBB bb = new AxisAlignedBB(mc.thePlayer.posX, mc.thePlayer.posY, mc.thePlayer.posZ, mc.thePlayer.posX, off, mc.thePlayer.posZ);
            if (!mc.theWorld.getCollidingBoundingBoxes(mc.thePlayer, bb).isEmpty()) {
                return true;
            }
        }
        return false;
    }

    private boolean inAir(final double height,final double plus){
        if (mc.thePlayer.posY < 0)
            return false;
        for (int off = 0; off < height; off += plus) {
            AxisAlignedBB bb = new AxisAlignedBB(mc.thePlayer.posX, mc.thePlayer.posY, mc.thePlayer.posZ, mc.thePlayer.posX, mc.thePlayer.posY - off, mc.thePlayer.posZ);
            if (!mc.theWorld.getCollidingBoundingBoxes(mc.thePlayer, bb).isEmpty()) {
                return true;
            }
        }
        return false;
    }
    
    @EventTarget
    public void onMotionUpdate(MotionEvent event) {
        if (!modeValue.get().equalsIgnoreCase("MLG"))
            return;

        if (!LiquidBounce.moduleManager.getModule(Fly.class).getState() && voidCheck.get() && inVoid()) return;

        if (event.getEventState() == EventState.PRE) {
            currentMlgRotation = null;
            currentMlgBlock = null;
            mlgTimer.update();

            if (!mlgTimer.hasTimePassed(10))
                return;

            if (mc.thePlayer.fallDistance > minFallDistance.get()) {
                FallingPlayer fallingPlayer = new FallingPlayer(
                        mc.thePlayer.posX,
                        mc.thePlayer.posY,
                        mc.thePlayer.posZ,
                        mc.thePlayer.motionX,
                        mc.thePlayer.motionY,
                        mc.thePlayer.motionZ,
                        mc.thePlayer.rotationYaw,
                        mc.thePlayer.moveStrafing,
                        mc.thePlayer.moveForward
                );

                double maxDist = mc.playerController.getBlockReachDistance() + 1.5;

                if (fallingPlayer.findCollision((int) Math.ceil((1.0 / mc.thePlayer.motionY) * (-maxDist))) == null)
                    return;
                
                currentMlgBlock = fallingPlayer.findCollision((int) Math.ceil((1.0 / mc.thePlayer.motionY) * (-maxDist)));

                boolean ok = new Vec3(mc.thePlayer.posX, mc.thePlayer.posY + mc.thePlayer.eyeHeight, mc.thePlayer.posZ).distanceTo(new Vec3(currentMlgBlock).addVector(0.5, 0.5, 0.5)) < mc.playerController.getBlockReachDistance() + Math.sqrt(0.75);

                if (mc.thePlayer.motionY < (currentMlgBlock.getY() + 1) - mc.thePlayer.posY) {
                    ok = true;
                }

                if (!ok)
                    return;

                int index = -1;

                for (int i = 36; i < 45; i++) {
                    ItemStack itemStack = mc.thePlayer.inventoryContainer.getSlot(i).getStack();

                    if (itemStack != null && (itemStack.getItem() == Items.water_bucket || itemStack.getItem() instanceof ItemBlock && ((ItemBlock) itemStack.getItem()).getBlock() == Blocks.web)) {
                        index = i - 36;

                        if (mc.thePlayer.inventory.currentItem == index)
                            break;
                    }
                }

                if (index == -1)
                    return;

                currentMlgItemIndex = index;
                

                if (mc.thePlayer.inventory.currentItem != index) {
                    PacketUtils.sendPacketNoEvent(new C09PacketHeldItemChange(index));
                }

                currentMlgRotation = RotationUtils.faceBlock(currentMlgBlock);
                currentMlgRotation.getRotation().toPlayer(mc.thePlayer);
            }
        } else if (currentMlgRotation != null) {
            ItemStack stack = mc.thePlayer.inventoryContainer.getSlot(currentMlgItemIndex + 36).getStack();

            if (stack.getItem() instanceof ItemBucket) {
                mc.playerController.sendUseItem(mc.thePlayer, mc.theWorld, stack);
            } else {
                Vec3i dirVec = EnumFacing.UP.getDirectionVec();

                if (mc.playerController.onPlayerRightClick(mc.thePlayer, mc.theWorld, stack, currentMlgBlock, EnumFacing.UP, new Vec3(dirVec.getX() * 0.5, dirVec.getY() * 0.5, dirVec.getZ() * 0.5).add(new Vec3(currentMlgBlock)))) {
                    mlgTimer.reset();
                }
            }

            if (mc.thePlayer.inventory.currentItem != currentMlgItemIndex)
                PacketUtils.sendPacketNoEvent(new C09PacketHeldItemChange(mc.thePlayer.inventory.currentItem));
        }
    }

    public void onDisable(){
        mc.timer.timerSpeed = 1.0F;
    }

    @Override
    public String getTag() {
        return modeValue.get();
    }
}

File: src/main/java/net/ccbluex/liquidbounce/features/module/modules/player/NoFall.kt
/*
 * FDPClient Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/SkidderMC/FDPClient/
 */
package net.ccbluex.liquidbounce.features.module.modules.player

import net.ccbluex.liquidbounce.config.Configurable
import net.ccbluex.liquidbounce.event.*
import net.ccbluex.liquidbounce.features.module.Category
import net.ccbluex.liquidbounce.features.module.Module
import net.ccbluex.liquidbounce.features.module.modules.modeNames
import net.ccbluex.liquidbounce.features.module.modules.selectedMode
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.aac.*
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.grim.Grim2371
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.matrix.*
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.normal.*
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.other.*
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.other.Blink
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.packet.Packet1
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.packet.Packet2
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.verus.Verus
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.vulcan.LatestVulcan
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.vulcan.OldVulcan
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.polar.Polar
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.sentinel.Sentinel
import net.ccbluex.liquidbounce.features.module.modules.player.nofallmodes.intave.Intave
import net.ccbluex.liquidbounce.utils.block.BlockUtils.collideBlock
import net.ccbluex.liquidbounce.utils.rotation.AlwaysRotationSettings
import net.ccbluex.liquidbounce.utils.rotation.RotationPriority
import net.minecraft.block.BlockLiquid
import net.minecraft.util.AxisAlignedBB.fromBounds
import net.minecraft.util.BlockPos
import net.minecraft.util.Vec3
import kotlin.math.max

object NoFall : Module("NoFall", Category.PLAYER, Category.SubCategory.PLAYER_COUNTER) {
    private val noFallModes = arrayOf(

        // Main
        SpoofGround,
        NoGround,
        Packet,
        Cancel,
        MLG,
        Blink,

        // Vanilla / Normal
        Vanilla,
        AlwaysSpoof,
        Damage,
        MotionFlag,
        Phase,

        // Packet
        Packet1,
        Packet2,

        // AAC
        AAC,
        LAAC,
        AAC3311,
        AAC3315,
        AAC44XFlag,
        AAC5014,
        AAC504,
        AACv4,

        // Hypixel (Watchdog)
        Hypixel,
        HypixelBlink,
        HypixelFlag,
        HypixelTimer,

        // Vulcan
        VulcanFast288,
        OldVulcan,
        LatestVulcan,

        // Matrix
        Matrix62x,
        Matrix62xPacket,
        MatrixCollide,
        Matrix663,
        MatrixNew,

        // Grim
        Grim2371,

        // Polar
        Polar,

        // Sentinel
        Sentinel,

        // Intave
        Intave,

        // Other Server
        Verus,
        Spartan,
        CubeCraft,
    )

    val mode by choices("Mode", noFallModes.modeNames(), "MLG")
        .describe("No-fall bypass method to use.")

    val minFallDistance by float("MinMLGHeight", 5f, 2f..50f) { mode == "MLG" }.subjective()
        .describe("Minimum fall height before MLG triggers.")

    val retrieveDelay: Int by int("RetrieveDelayTicks", 5, 1..10) {
        mode == "MLG"
    }.onChanged {
        maxRetrievalWaitingTimeValue.set(max(maxRetrievalWaitingTime, it))
    }.subjective()
        .describe("Ticks to wait before retrieving the placed block.")

    private val maxRetrievalWaitingTimeValue = int("MaxRetrievalWaitingTime", 10, 1..20) {
        mode == "MLG"
    }.onChange { _, new ->
        new.coerceAtLeast(retrieveDelay)
    }
        .describe("Maximum ticks to wait while retrieving the block.")

    val maxRetrievalWaitingTime by maxRetrievalWaitingTimeValue

    val autoMLG by choices("AutoMLG", arrayOf("Off", "Pick", "Spoof"), "Spoof") { mode == "MLG" }
        .describe("How to place a block to save yourself.")
    val swing by boolean("Swing", true) { mode == "MLG" }
        .describe("Swing the arm when placing the MLG block.")

    val options = AlwaysRotationSettings(this) { mode == "MLG" }
        .withRequestPriority(RotationPriority.CRITICAL)
    val matrixSafe by boolean("SafeNoFall", true) { mode == "Matrix6.6.3" }
        .describe("Use the safer Matrix no-fall variant.")
    val motionFlagSpeed by float("MotionFlag-MotionSpeed", -0.01f, -5f..5f) { mode == "MotionFlag" }
        .describe("Vertical motion sent in MotionFlag mode.")
    val phaseOffset by int("PhaseOffset", 1, 0..5) { mode == "Phase" }
        .describe("Offset used in Phase no-fall mode.")
    val hypixelBlinkIndicator by boolean("Indicator", true) { mode == "HypixelBlink" }
        .describe("Show an indicator in HypixelBlink mode.")

    // Using too many times of simulatePlayer could result timer flag. Hence, why this is disabled by default.
    val checkFallDist by boolean("CheckFallDistance", false) { mode == "Blink" }.subjective()
        .describe("Only blink within a fall-distance range.")
    val fallDist by floatRange("FallDistance", 2.5f..20f, 0f..100f) {
        mode == "Blink" && checkFallDist
    }.subjective()
        .describe("Fall-distance range within which to blink.")

    val autoOff by boolean("AutoOff", true) { mode == "Blink" }
        .describe("Disable Blink no-fall after landing.")
    val simulateDebug by boolean("SimulationDebug", false) { mode == "Blink" }.subjective()
        .describe("Render fall simulation debug visuals.")
    val fakePlayer by boolean("FakePlayer", true) { mode == "Blink" }.subjective()
        .describe("Spawn a fake player while blinking.")

    private val mlgGroup = Configurable("MLG")
    private val blinkGroup = Configurable("Blink")
    private val variantsGroup = Configurable("Variants")

    init {
        options.nestInto(mlgGroup)
        moveValues(mlgGroup,
            "Mode", "MinMLGHeight", "RetrieveDelayTicks", "MaxRetrievalWaitingTime", "AutoMLG", "Swing")
        moveValues(blinkGroup,
            "CheckFallDistance", "FallDistance", "AutoOff", "SimulationDebug", "FakePlayer")
        moveValues(variantsGroup,
            "SafeNoFall", "MotionFlag-MotionSpeed", "PhaseOffset", "Indicator")

        addValues(listOf(mlgGroup, blinkGroup, variantsGroup))
    }
    var currentMlgBlock: BlockPos? = null
    var retrievingPos: Vec3? = null
    var wasTimer = false

    override fun onEnable() {
        modeModule.onEnable()
        retrievingPos = null
    }

    override fun onDisable() {
        if (mode == "MLG") {
            currentMlgBlock = null
            retrievingPos = null
        }

        if (wasTimer) {
            mc.timer.timerSpeed = 1f
            wasTimer = false
        }

        modeModule.onDisable()
    }

    val onTick = handler<GameTickEvent> {
        modeModule.onTick()
    }

    val onUpdate = handler<UpdateEvent> {
        val thePlayer = mc.thePlayer

        if (wasTimer) {
            mc.timer.timerSpeed = 1f
            wasTimer = false
        }

        if (collideBlock(thePlayer.entityBoundingBox) { it is BlockLiquid } || collideBlock(
                fromBounds(
                    thePlayer.entityBoundingBox.maxX,
                    thePlayer.entityBoundingBox.maxY,
                    thePlayer.entityBoundingBox.maxZ,
                    thePlayer.entityBoundingBox.minX,
                    thePlayer.entityBoundingBox.minY - 0.01,
                    thePlayer.entityBoundingBox.minZ
                )
            ) { it is BlockLiquid }
        ) return@handler

        modeModule.onUpdate()
    }

    val onRender3D = handler<Render3DEvent> {
        modeModule.onRender3D(it)
    }

    val onRender2D = handler<Render2DEvent> {
        modeModule.onRender2D(it)
    }

    val onPacket = handler<PacketEvent> {
        mc.thePlayer ?: return@handler

        modeModule.onPacket(it)
    }

    val onBB = handler<BlockBBEvent> {
        mc.thePlayer ?: return@handler

        modeModule.onBB(it)
    }

    // Ignore condition used in LAAC mode
    val onJump = handler<JumpEvent>(always = true) {
        modeModule.onJump(it)
    }

    val onStep = handler<StepEvent> {
        modeModule.onStep(it)
    }

    val onMotion = handler<MotionEvent> {
        modeModule.onMotion(it)
    }

    val onMove = handler<MoveEvent> {
        val thePlayer = mc.thePlayer

        if (collideBlock(thePlayer.entityBoundingBox) { it is BlockLiquid }
            || collideBlock(
                fromBounds(
                    thePlayer.entityBoundingBox.maxX,
                    thePlayer.entityBoundingBox.maxY,
                    thePlayer.entityBoundingBox.maxZ,
                    thePlayer.entityBoundingBox.minX,
                    thePlayer.entityBoundingBox.minY - 0.01,
                    thePlayer.entityBoundingBox.minZ
                )
            ) { it is BlockLiquid }
        ) return@handler

        modeModule.onMove(it)
    }

    val onRotationUpdate = handler<RotationUpdateEvent> {
        modeModule.onRotationUpdate()
    }

    override val tag
        get() = mode

    private val modeModule
        get() = noFallModes.selectedMode(mode)
}
